### PR TITLE
fix(otp-header): persist otp header on subsequent requests

### PR DESCRIPTION
**Pull request by ruxandrafed**
Related to #1279. Adds test to check for presence of otp header in requests made to list authorizations.

### CODE REVIEW

**Comment by gr2m** [comment]
You have to add a `-test` suffix to the test in order for it to be run by the `npm test` script. Other than that the test looks 💯👍

**Comment by ruxandrafed** [comment]
@gr2m Thanks for catching that! Fixed.

**Comment by gr2m** [comment]
Authentication works using an internal plugin, it exists here:
https://github.com/octokit/rest.js/tree/master/plugins/authentication

plugins can hook into the request lifecycle using `octokit.hook.*`, in this case the authentication plugin runs code before a request is sent and in case of an error:

https://github.com/octokit/rest.js/blob/21e7a017366a3b029413d5ad51f34741bb0b59b8/plugins/authentication/index.js#L19-L20

In the error handler, we check if the error was due to a missing OTP, in which case we run your `auth.on2fa()` method and retry the request with the additional header

https://github.com/octokit/rest.js/blob/21e7a017366a3b029413d5ad51f34741bb0b59b8/plugins/authentication/request-error.js#L22-L31

Now what we want is not only to retry this request but also set the `x-github-otp` header for subsequent requests. 

So what I would probably try to do is use the `state` variable that is passed implicitly per pre-binding to both the before- and error-handler and store the otp on it. Then in the before handler check if it’s set and if it is, set the `x-github-otp` header.

That would be a good start I think :)

**Comment by ruxandrafed** [comment]
Thanks @gr2m 😄 I actually just left a comment in the issue before I saw your reply. Pushed one more commit with the fix, it should pass the test, but I look forward to your review.

File: plugins/authentication/before-request.js [hunk 1]
```diff
@@ -3,6 +3,10 @@ module.exports = authenticationBeforeRequest
 const btoa = require('btoa-lite')
 
 function authenticationBeforeRequest (state, options) {
+  if (state.otp) {
+    options.headers['x-github-otp'] = state.otp
+  }
+
```
Comment (gr2m) [inline plugins/authentication/before-request.js]: maybe check for `state.otp` below after we check for `state.auth.username`, as it’s only relevant in combination with Basic authentication?

File: plugins/authentication/request-error.js [hunk 2]
```diff
@@ -27,6 +27,7 @@ function authenticationRequestError (state, error, options) {
       const newOptions = Object.assign(options, {
         headers: Object.assign({ 'x-github-otp': oneTimePassword }, options.headers)
       })
+      state.otp = oneTimePassword
```
Comment (gr2m) [inline plugins/authentication/request-error.js:27]: 👍

**Review by gr2m — COMMENTED** [review]
Looks good, just one comment

File: plugins/authentication/before-request.js (hunk 1 above)
Comment (ruxandrafed) [reply plugins/authentication/before-request.js]: Makes sense! Moved it below: https://github.com/octokit/rest.js/pull/1281/commits/856a7912786a75f7a9cf1b440cfd43ea5ac6b41e.

**Comment by gr2m** [comment]
I wonder if we should update this condition here

https://github.com/octokit/rest.js/blob/21e7a017366a3b029413d5ad51f34741bb0b59b8/plugins/authentication/request-error.js#L14

It will fail if the set OTP becomes invalid. I think the better behaviour would be that it would ask for an OTP again as soon as it becomes invalid, but only if it was valid before. We could additionally check if `state.otp` is set? We would also need to unset it so we don’t up in an infinite loop the user keeps entering an invalid code.

Maybe something like this?

```js
  if (error.status === 401 && otpRequired && error.request && error.request.headers['x-github-otp']) {
    if (state.otp) {
      delete state.otp // no longer valid, request again
    } else {
      throw new HttpError('Invalid one-time password for two-factor authentication', 401, error.headers, options)
    }
  }
```

**Comment by gr2m** [comment]
^^ we can also do that in a follow up PR if you prefer, the current PR looks good as is

**Comment by ruxandrafed** [comment]
@gr2m Sounds good, I'll do another PR! Thank you.

File: plugins/authentication/request-error.js [hunk 3]
```diff
@@ -27,6 +27,7 @@ function authenticationRequestError (state, error, options) {
...
       return state.octokit.request(newOptions)
```
Comment (gr2m) [inline plugins/authentication/request-error.js]: I _think_ that with the addition to `before-request.js`, we no longer need to set the `x-github-otp` header here, we could remove lines 27-29 and just do `return state.octokit.request(options)`. I didn’t test it though.

I’ll be offline now and back tomorrow

Comment (ruxandrafed) [reply plugins/authentication/request-error.js]: You're right, it makes sense. Refactored in https://github.com/octokit/rest.js/pull/1281/commits/e635dc9d4b2df22ccd0e25f61a6eb9703a255854.

Will look at the other changes tomorrow. Have a good night!

**Review by gr2m — APPROVED** [review]
A+ pull request would merge again 👍 Thanks!

**Pull request merged**